carsus/io/cmfgen/cmfgen_.py

```diff
@@ -0,0 +1,328 @@
+
+        try:
+            n = find_row(fname, "Number of transitions").split()[0]
+            kwargs['nrows'] = int(n)
+    
+        except AttributeError:
+            pass
+    
+        try:
+            names = find_row(fname, 'ransition\T').split()
+            names = [ n.replace('D', 'E') for n in names ]
+            kwargs['names'] = ['State A', 'State B'] + names[1:]
+
+        except AttributeError:
+            warnings.warn('No column names')  # TODO: some files have no column names nor header
+    
+        try:
+            df = pd.read_csv(fname, **kwargs, engine='python')
+            df.iloc[:,2:] = df.iloc[:,2:].multiply(1e+04)  # FIXME: this apply for all COL files?
+    
+        except pd.errors.EmptyDataError:
+            df = pd.DataFrame()
+            warnings.warn('Empty table')
+
+        self.base = df
+        self.columns = df.columns.tolist()
+
+
+class CMFGENPhotoionizationCrossSectionParser(BaseParser):
+    """
+        Description
+        ----------
+        base : list of pandas.DataFrame 's
+        columns : dictionary with photoionization cross-section file metadata
+            
+        Methods
+        -------
+        load(fname)
+            Parses the input data and stores the results in the `base` attribute
+    """
+    keys = ['!Date',  # Metadata to parse from header
+            '!Number of energy levels', 
+            '!Number of photoionization routes', 
+            '!Screened nuclear charge',
+            '!Final state in ion',
+            '!Excitation energy of final state',
+            '!Statistical weight of ion',
+            '!Cross-section unit',
+            '!Split J levels',
+            '!Total number of data pairs',
+            ]
+
+    def _table_gen(self, f):
+        """ Generator. Yields a cross section table for an energy level """
+        meta = {}
+        data = []
+
+        for line in f:
+
+            if '!Configuration name' in line:
+                meta['Configuration'] = line.split()[0]
+
+            if '!Type of cross-section' in line:
+                meta['Type of cross-section'] = line.split()[0]
+
+            if '!Number of cross-section points' in line:
+                meta['Points'] = int(line.split()[0])
+
+                for i in range(meta['Points']):
+
+                    values = f.readline().split()
+                    if len(values) > 2:  # Verner ground state fits case
+
+                        data.append( list(map(int, values[:2])) + list(map(float, values[2:])) )
+
+                        if i == meta['Points']/len(values) -1:
+                            break
+
+                    else:
+                        data.append(map(to_float, values))              
+
+                break
+
+        df = pd.DataFrame.from_records(data)
+        df._meta = meta    
+
+        yield df
+
+
+    def load(self, fname):
+    
+        meta = {}
+        tables = []
```

Comment:
Do you think it would be a good idea to use an OrderedDict with the Configurations as the keys instead of a list?

Reply:
If we stick to the list format, then the DataFrames are ordered by the transition number assigned by the CMFGEN people. This was very convenient for me at this stage. I don't know what will be more useful for you in the future.

Anyways, when dumping to HDF we can choose to assign group names after the configuration.

It's up to you!

Reply:
If we stick to the list format, then the DataFrames are ordered by the transition number assigned by the CMFGEN people. This was very convenient for me at this stage. I don't know what will be more useful for you in the future.

Anyways, when dumping to HDF we can choose to assign group names after the configuration.

It's up to you!